gkr_iop/src/circuit_builder.rs

```diff
@@ -1073,25 +1092,26 @@ impl<'a, E: ExtensionField> CircuitBuilder<'a, E> {
             let mut rep_x = rep_x.to_owned();
             rep_x.rotate_right(chunks_rotation);
 
-            for i in 0..2 {
-                // The respective 4 elements in the byte representation
-                let lhs = rep8[4 * i..4 * (i + 1)]
+            // 64 bits represent in 4 limb, each with 16 bits
+            let num_limbs = 4;
+            let mut rep_x_iter = rep_x.iter().cloned();
+            for limb_i in 0..num_limbs {
+                let lhs = rep8[2 * limb_i..2 * (limb_i + 1)]
                     .iter()
                     .map(|wit| (8, wit.expr()))
                     .collect_vec();
-                let cnt = rep_x.len() / 2;
-                let rhs = &rep_x[cnt * i..cnt * (i + 1)];
+                let rhs_limbs = take_til_threshold(&mut rep_x_iter, 16, &|limb| limb.0).unwrap();
 
-                assert_eq!(rhs.iter().map(|e| e.0).sum::<usize>(), 32);
+                assert_eq!(rhs_limbs.iter().map(|e| e.0).sum::<usize>(), 16);
 
-                self.require_reps_equal::<32, _, _>(
+                self.require_reps_equal::<16, _, _>(
```

Comment:
key highlights: constrain on 16 bits limb individually instead of 32 bits (limb_16 + (1 << 16( * limb_16) because later are larger than some finite field characteristic, e.g. babybear